Material: Add "material.alphaHash" transparency mode

donmccurdy opened the pull request:
**Related issues:**
 
- https://github.com/mrdoob/three.js/issues/9977
- https://github.com/mrdoob/three.js/issues/19554

**Description:**

Proposal for adding `.alphaHash: boolean` and `.alphaHashScale: number` properties to all three.js Material types. This is an alternative method of transparency distinct from alpha blending (`.transparent=true`) and physically-based transmission (`.transmission > 0`), which is not a panacea but has some significant benefits:

1. Keeps objects in the 'opaque' render list
2. Avoids all sorting issues associated with alpha blending
3. Less overdraw than alpha blending
4. Less performance cost than transmission
5. Keeps objects visible through transmissive materials

The major drawback of the method is noise / grain, which can be reduced significantly with appropriate anti-aliasing. With or without AA, alpha hash is a good choice in several use cases:

1. Transitioning visibility of opaque objects without triggering sorting and backface problems during the transition. E.g. switching between LODs should not cause distracting flicker.
2. Hiding boundaries between two opaque objects, e.g. blades of grass fade into the ground, without requiring everything to be transparent.
3. Scenes with so much overlapping transparency that sorting issues are intractable

In these cases a bit of grain might be a reasonable tradeoff to avoid alpha blending issues — particularly if the transparent region is temporary, small, or distant from the camera.

The effect closely resembles Blender's *Blend Mode: Alpha Hashed* material property. Blender enables multi-sampling by default, which hides most of the grain, but this can be disabled in its renderer settings. The `.alphaHashScale: number` parameter should be used to tune the hash function to fit the multi-sampling method, with values in the range of 0.1–0.5 working well in my tests. When multi-sampling is not used, 1.0 is fine.

| # samples | three.js | blender |
|-----------|----------|---------|
| 1 | <img width="557" alt="three_1" src="https://user-images.githubusercontent.com/1848368/175824634-e76ccfbe-14cb-4119-aa75-e3e5be323f91.png"> | <img width="544" alt="blender_1" src="https://user-images.githubusercontent.com/1848368/175825142-397ce964-f9d7-4609-9ecd-c0a9a70c8cc8.png"> |
| 16 | <img width="576" alt="three_16" src="https://user-images.githubusercontent.com/1848368/175824633-aeee1685-329f-46cf-aac9-8b84659f7273.png"> | <img width="589" alt="blender_16" src="https://user-images.githubusercontent.com/1848368/175825145-b8ae06c8-5652-43fd-bda0-76c3531026d8.png"> |






**Questions:**

- [x] ~~Do we want to support `material.alphaHash`? We could also consider a NodeMaterial-only solution, as a separate node: `material.alphaTest = new AlphaHashNode( opacityNode, scale );`~~
- [x] ~~TAARenderPass is brightening the scene, and I'm not sure why. I've omitted all color management settings for now, to isolate that underlying problem. Any ideas here? Is this an appropriate type of multi-sampling to be using for this example? /cc @bhouston~~ See below.
- [x] ~~Better example? I used this [chloroplast scene](https://sketchfab.com/3d-models/chloroplast-42eec3ca7c074048b57f94ea9db7e7a8) from Sketchfab, mainly because it doesn't work well with alpha blending, for any choice of depthWrite.~~

**References:**

- https://casual-effects.com/research/Wyman2017Hashed/index.html

***

*This contribution is funded by [The New York Times](https://www.nytimes.com/).*

donmccurdy committed:
Material: Add .alphaHash and .alphaHashScale

Material.alphaHash: Add example

Clean up.

Update builds

Examples: Clean up alpha hash params

Fix background color changes, add GammaCorrectionShader

TAARenderPass: Initialize .accumulateIndex

Alpha Hash: Initialize sampleLevel=0 in example

Avoids flashing from TAARenderPass

Clean up alpha hash example.

Lint

Update builds

Add screenshot

bhouston commented:
This is really cool!  It is fast and the results are nice!

With regards to TAA causing things to get brighter when it is enabled, it appears to do this with or without your changes.  I checked here and it has that same bug in main: https://threejs.org/examples/?q=taa#webgl_postprocessing_taa. This wasn't the case a long time ago so some change in the last year or two must have broken it.  But it is unrelated to your changes.

donmccurdy commented:
Thanks @bhouston! I'll file that as a separate issue then, I'm having some other trouble with color management in post-processing (unrelated to TAARenderPass) that may be worth looking at.

Assuming we'll find a fix so that the scene brightness doesn't fluctuate, I would be interested in everyone's feedback on whether we want to support this. Happy to clean up the example and implementation more if so. :)

Mugen87 commented:
> With regards to TAA causing things to get brighter when it is enabled,

@donmccurdy Can you please check if reverting #23671 fixes the issue? If so, `SSAARenderPass` might need a custom copy pass.

donmccurdy commented:
Thanks @Mugen87! Unfortunately I don't see a difference after reverting #23671. The same thing happens in the [official TAA example](https://threejs.org/examples/?q=taa#webgl_postprocessing_taa) if I add this line:

```js
scene.background = new THREE.Color( 0x505050 );
```

donmccurdy commented:
Possibly https://github.com/mrdoob/three.js/issues/9051?

LeviPesin commented on examples/webgl_materials_alphahash.html:
```diff
@@ -0,0 +1,236 @@
+			import { RenderPass } from './jsm/postprocessing/RenderPass.js';
+			import { ShaderPass } from './jsm/postprocessing/ShaderPass.js';
+			import { TAARenderPass } from './jsm/postprocessing/TAARenderPass.js';
+			import { GammaCorrectionShader } from './jsm/shaders/GammaCorrectionShader.js';
```
`GammaCorrectionShader` is unused.

donmccurdy replied:
Lots to clean up here – I'll do cleanup if there's interest in adding the feature.

donmccurdy commented:
Setting `scene.background` explicitly to black fixes the most noticeable flashing in the demo, although the model itself still shows a bit of flicker caused by the brightening issue I think.  Hopefully that is good enough to evaluate and consider the `.alphaHash` feature for now.

hybridherbst commented:
Great to see! Typically transparency-through-noise can be improved with alpha to coverage, have you tried that? Could reduce the noise even more for low-ish AA values.

donmccurdy commented:
Good point — to be honest I haven't really used alpha-to-coverage before. Assuming I've set that up right (alphaHash=true, alphaToCoverage=true, transparent=false, renderer.antialias=true) the difference is visible, but not in a way that seems obviously better or worse in this example:

| alphaToCoverage=true | alphaToCoverage=false |
|---|---|
| <img width="738" alt="atc_on" src="https://user-images.githubusercontent.com/1848368/176204837-7e6c7424-aaea-452e-8539-53cfb8feb3a6.png"> | <img width="738" alt="atc_off" src="https://user-images.githubusercontent.com/1848368/176204825-b4e605b8-eeec-4c72-af17-6f3396f76902.png"> |

Despite the checkbox in the screenshot, TAA is off for both. I'm not confident enough about setting up post-processing to combine TAA with WebGL2 anti-aliased render targets. So there's more grain than in the original screenshots — there are probably better ways to set this up and minimize noise.

donmccurdy commented:
@gkjohnson from the paper, emphasis mine —

> User parameter `g_HashScale` controls the target noise scale in pixels (default 1.0).
Changing g_HashScale is useful if the chosen hash outputs noise at
another frequency. Also, _**when temporal antialiasing, using noise
below pixel scale (e.g., 0.3–0.5) allows for temporal averaging**_.

With TAA, I'm finding 0.1 looks pretty good. This may depend on the number of samples. Without TAA, I also see an improvement when decreasing the scale from 1 to 0.5, but not much change below that — unsure why that would happen.

The paper also discusses "Applications to Alpha-to-Coverage" but I haven't gotten into that area yet, it may require more work.

donmccurdy commented:
Hmm, from one of the authors...

> To move to alpha to coverage, you need to independently compute the binary visibility tests for each sample and explicitly output them, e.g., via gl_SampleMask.  You cannot output an alpha value from the shader and let the hardware apply its default dithering.

https://twitter.com/_cwyman_/status/935743334244605953

I don't think WebGL 1.0 or 2.0 include `gl_SampleMask`, that's an OpenGL 4.0 feature. :/

donmccurdy commented:
I've filed the issues about TAARenderPass at https://github.com/mrdoob/three.js/issues/24318 — it turns out we can work around them both by using a black scene background and keeping `pass.sampleLevel = 0`. Because `gl_SampleMask` is not available, alpha-to-coverage and multi-sampling can't reduce the noise associated with Hashed Alpha Testing, and so TAA seems to be the best option for reducing that noise. On higher-DPI displays or smaller surfaces, the noise is much less noticable and TAA may not be needed. 

Here's the preview link once again:

- [webgl / materials / alphahash](https://raw.githack.com/donmccurdy/three.js/feat-alphahash/examples/index.html?q=alpha#webgl_materials_alphahash)

Visually, this is comparable to Blender, so I probably won't make any further changes here unless we decide to move forward with the feature. If so, some further testing and cleanup would be needed.

sdarpeng commented:
You should have this model to process your branch test.
@donmccurdy 
[hair_test1.zip](https://github.com/mrdoob/three.js/files/9972170/hair_test1.zip)

donmccurdy commented:
I believe this is a useful improvement for the example shared by @sdarpeng as well. The appearance is stable (no sort issues) under different camera angles.

| before | after |
|---|---|
| ![Screenshot 2023-06-07 at 10 11 06 AM](https://github.com/mrdoob/three.js/assets/1848368/69b19ffa-2e6f-4fc2-be96-f64f155f8a88) | ![Screenshot 2023-06-07 at 10 11 04 AM](https://github.com/mrdoob/three.js/assets/1848368/62669127-90fc-4801-afb7-02e59a14e345) |

WestLangley commented on src/renderers/shaders/ShaderLib/meshphysical.glsl.js:
```diff
@@ -52,6 +58,12 @@ void main() {
+
+#ifdef USE_ALPHAHASH
+
+	vObjectPosition = transformed.xyz;
```
Can the existing varying `vWorldPosition` be used, instead?

Also, what is the reason this is chunk is physical-material-specific?

donmccurdy replied:
Yes, and only for convenience of demonstrating the feature. I'll add it to other material types if we think this PR might move forward!

donmccurdy replied:
@WestLangley Correction – `vWorldPosition` does not work as well, for reasons mentioned in the paper and apparent after I tested this:

> Hashing world-space coordinates provides stable noise for static geometry, and our early tests used world-space coordinates. However, this fails on dynamic geometry. Object-space coordinates give stable hashes for skinned and rigid transforms and dynamic cameras.

Given that we want the hash anchored to skinned and morphed surfaces, perhaps vObjectPosition is not the right name? We simply want to write the `position` attribute to a varying, would `vPosition` be appropriate?

WestLangley replied:
Yes.

Mugen87 commented:
> Do we want to support material.alphaHash? We could also consider a NodeMaterial-only solution, as a separate node: material.alphaTest = new AlphaHashNode( opacityNode, scale );

I vote to add `Material.alphaHash`. In this way, we can provide at least one way in the core for addressing typical transparency artifacts related to alpha blending. 

IMO, TAA _should_ be used for acceptable image quality. But given the manageable amount of changes related to the core, I think it's worth an addition.

mrdoob commented:
Yeah, this looks good to me 👍

As per the example, maybe something like [this one](https://threejs.org/examples/#webgl_instancing_raycast) but every sphere has a different color and alpha `0.5`?

donmccurdy commented:
> As per the example, maybe something like [this one](https://threejs.org/examples/#webgl_instancing_raycast) but every sphere has a different color and alpha 0.5?

I might shuffle the order of instances to make the problems with alpha blending more obvious from any angle, but yes that sounds fine!

If there are no objections, I'd like to remove the `.alphaHashScale` property for now, using a hard-coded default instead. I haven't found any way to create a scene that benefits from changing its value, despite my earlier response, so including the property feels premature.

donmccurdy committed:
Material: Move alphahash to separate shader chunks.

donmccurdy committed:
Material: Remove alphaHashScale

donmccurdy committed:
Clean up

donmccurdy committed:
Clean up

donmccurdy committed:
Clean up

donmccurdy commented:
To simplify things, I've removed the demo from this PR, and will create a new demo in a separate PR.

I've also moved the _alphaHash_ GLSL out of the _alphaTest_ shader chunks, and into its own chunks. Happy to revert that if we'd prefer otherwise.

donmccurdy committed:
Clean up

donmccurdy commented:
PR for the example:

- https://github.com/mrdoob/three.js/pull/26341

sdarpeng commented:
Greaaaaat Job, guys.

sdarpeng commented:
https://github.com/mrdoob/three.js/assets/55733077/f7c09e5c-7d67-4e85-9795-a6a74fbfe8d6

The new three.js alphaHash transparency looks fucking good.
